src/backend/columnar/columnar_customscan.c

```diff
@@ -679,19 +705,129 @@ ExprReferencesRelid(Expr *expr, Index relid)
  * and that doesn't seem worth the effort. Here we just look for "Var op Expr"
  * or "Expr op Var", where Var references rel and Expr references other rels
  * (or no rels at all).
+ *
+ * Moreover, this function also looks into BoolExpr's to recursively extract
+ * pushdownable OpExpr's of them:
+ * i)   AND_EXPR:
+ *      Take pushdownable args of AND expressions by ignoring the other args.
+ * ii)  OR_EXPR:
+ *      Ignore the whole OR expression if we cannot exract a pushdownable Expr
+ *      from one of its args.
+ * iii) NOT_EXPR:
+ *      Simply ignore NOT expressions since we don't expect to see them before
+ *      an expression that we can pushdown, see the comment in function.
+ *
+ * For example, given following WHERE clause:
+ * (
+ *     (a > random() OR a < 30)
+ *     AND
+ *     a < 200
+ * ) OR
+ * (
+ *     a = 300
+ *     OR
+ *     a > 400
+ * );
```

Comment:
To maintain sanity, we should not try to handle any form other than CNF (https://en.wikipedia.org/wiki/Conjunctive_normal_form). If we do that, then each conjunct will just be a list of ORs, and then we can reasonably deal with it at execution time.

We can either turn it into CNF (if we find some infrastructure for doing that), or just make it only work on CNF and ignore more complex formulations (perhaps emitting a debug message).

If we don't use CNF, it will be harder to reason about, and it will also be harder to ever come up with a reasonable way to execute that doesn't depend on `predicate_refuted_by()`.

Reply:
I think we can use `canonicalize_qual` for that